tools/gen_data.py

```diff
@@ -0,0 +1,272 @@
+from __future__ import (
+    division, 
+    print_function
+)
+
+import base64
+import logging
+import numpy as np
+import os
+import string
+import sys
+if sys.version_info[0] == 3:
+    _get_byte = lambda c: c
+else:
+    _get_byte = ord
+import time
+
+from argparse import ArgumentParser
+from collections import defaultdict
+from functools import partial
+from google.protobuf.internal.encoder import _VarintBytes
+from multiprocessing import Pool
+from proto.offline_data_pb2 import Datum
+
+ALPHABET = np.array(list(string.ascii_lowercase + string.digits))
+MULTIPLIERS = {
+    "b" : 1024 ** 0,
+    "kb": 1024 ** 1,
+    "mb": 1024 ** 2,
+    "gb": 1024 ** 3,
+}
+
+KEY_SIZE = 12   # 8 bytes + overhead from base64 encoding
+MASTER_SIZE = 2
+FILE_EXTENSION = '.dat'
+
+logging.basicConfig(
+    level=logging.INFO,
+    format='%(asctime)s - %(process)d - %(levelname)s: %(message)s'
+)
+LOG = logging.getLogger("gen_data")
+LOG_EVERY_SEC = 1
+
+def encode_key(key: int) -> bytes:
+    """
+    Encodes an integer key into a fixed length string
+    """
+    return base64.b64encode(
+        key.to_bytes(8, byteorder='little')
+    )
+
+
+def fnv_hash(value: bytes) -> int:
+    assert isinstance(value, bytes)
+
+    FNV_32_PRIME = 0x01000193
+    FNV_32_SIZE = 2**32
+
+    hash = 0x811c9dc5
+    for byte in value:
+        hash = (hash * FNV_32_PRIME) % FNV_32_SIZE
+        hash = hash ^ _get_byte(byte)
+    return hash
+
+
+class DataGenerator:
+
+    def __init__(
+        self,
+        data_dir: str,
+        prefix: str,
+        num_replicas: int,
+        num_partitions: int,
+        size: int,
+        size_unit: str,
+        record_size: int,
+        max_jobs: int,
+    ):
+        self.data_dir = os.path.abspath(data_dir)
+        self.prefix = prefix
+        self.num_partitions = num_partitions
+        self.num_replicas = num_replicas
+
+        total_size = MULTIPLIERS[size_unit] * size
+        num_records = total_size / (KEY_SIZE + record_size + MASTER_SIZE)
+        self.num_records = int(num_records)
+        self.record_size = record_size
+        self.max_jobs = max_jobs
+
+    def key_to_partition(self, key: int) -> int:
+        return fnv_hash(encode_key(key)) % self.num_partitions
+
+    def gen_data(self, as_text) -> None:
+        start_time = time.time()
+
+        # Identify the keys of each partition
+        LOG.info(
+            "Creating %d partitions from %d keys...", 
+            self.num_partitions, self.num_records
+        )
+        key_per_part = defaultdict(list)
+        for key in range(0, self.num_records):
```

Comment:
Might be worth documenting that these will be the keys. It was a little confusing, mostly bc I'm unfamiliar with the partial function syntax.

Also could point out that the random seed only affects the data values, not the keys